### PR TITLE
Fix stale cache reads on equal-size, same-mtime transcript rewrites

### LINKED ISSUE #560: transcript parse cache can serve stale records on an equal-size in-place rewrite
**Severity:** Low
**Model:** haiku

Where: tokenjam/ (sub-repo) — tokenjam/core/optimize/transcript_cache.py
Observed: Cache validity is keyed only on (path, size, mtime). An in-place equal-length rewrite that preserves mtime, or two same-size rewrites inside one filesystem mtime tick, would serve stale cached records. Low real-world risk today because Claude Code transcripts are append-only JSONL that grow monotonically, so worst case is one-tick-delayed invalidation rather than permanent staleness.
Fix direction: add a cheap content fingerprint (hash of first/last N bytes) alongside size+mtime as a belt-and-suspenders check.
Done when: an equal-size in-place rewrite invalidates the cache, covered by a test.
Ref: raised by the review bot on the shared-helpers PR (public repo PR 511).

### PR DESCRIPTION
**Pull request by Vam-si-krish**
The transcript cache previously trusted `(size, mtime)`, which can miss equal-size in-place rewrites. This adds a full-content fingerprint while keeping cache misses and concurrent rewrites safe.

## Summary

- Stream SHA-256 over the complete transcript so middle-only rewrites invalidate the cache.
- Compute the fingerprint lazily only when a loaded entry already matches `size` and `mtime`.
- Re-stat and re-fingerprint after parsing; skip the cache write when the source changed during the parse.
- Preserve deliberate `None` fingerprints for unreadable-but-stattable files while invalidating both readable-to-unreadable and unreadable-to-readable transitions.
- Document that cache hits still read the source bytes for hashing but avoid the more expensive JSONL parse.

## Issue #560: equal-size transcript rewrites

The original metadata-only validity check could return stale parsed records when content changed without changing file size or modification time. A streaming whole-file digest closes that gap, including changes located away from the head and tail of large transcripts.

The revalidation step also prevents a digest observed before parsing from being stored beside records produced from different content. If the source moves while parsing, the fresh records are returned to the caller but no internally inconsistent cache entry is persisted.

## Tests / Verification

- `.venv/bin/python -m pytest tests/unit/test_transcript_cache.py -q` — 18 passed
- `.venv/bin/ruff check tokenjam/` — passed
- `.venv/bin/mypy tokenjam/` — 268 source files checked, no issues
- Full non-e2e suite — 5,062 passed, 7 skipped
- Six unrelated quickstart/summarize failures in the full run were reproduced identically on a clean `origin/main` worktree; none touch the transcript cache.

## What's NOT in this PR

- No changes to transcript parsing or the opt-in cache call sites.
- No attempt to fix the six unrelated failures currently present on `main`.
- No bounded head/tail sampling; every byte is intentionally hashed so middle-only rewrites are visible.

Closes #560

🤖 Generated with OpenAI Codex

Co-Authored-By: anshss <noreply@github.com>

### CODE REVIEW

**Comment by anshss** [comment]
@Vam-si-krish Thanks for this pr bro! A few things before merge:

1. **Make the fingerprint lazy.** It's computed unconditionally before `_load`, so
   every cache miss hashes the whole file and then `_parse_records` immediately reads
   it again. Move it inside the hit check, after `size` and `mtime` already match.

2. **Close the store-time gap.** The fingerprint is taken before `_parse_records`, so
   a rewrite landing between the two stores one version's fingerprint against another
   version's records. Re-stat and re-fingerprint after the parse, and skip the store if
   anything moved.

3. **Fix `test_unreadable_file_still_reuses_its_cache_entry`.** It patches `_fingerprint`
   to `None` at warm time as well, so it never exercises the real
   warm-then-becomes-unreadable transition it's named for.

4. **Update the PR body.** It still describes the 4096-byte head/tail design and its I/O
   bound, which the current code no longer does. Please also drop the line about the
   suite failing on the upstream base: it's green at `491d13f7` (3643 passed, 1 skipped).

5. **Note the tradeoff in the module docstring.** Hashing every byte means a cache hit
   now skips the parse but no longer skips the read. Worth stating explicitly so the
   next reader doesn't re-litigate it.

**Comment by anilmurty** [comment]
welcome and thanks for the contribution @Vam-si-krish !

**Comment by anilmurty** [comment]
Adding to @anshss's review (his 5 points are the right gate — lazy fingerprint, the store-time divergence gap, the unreadable-file test, the PR body, and the docstring tradeoff all stand). Two things his list doesn't cover:

1. **CI hasn't actually run yet.** The Python test matrix (`test (3.10/3.11/3.12)`) is still pending — this is an external-fork PR, so a maintainer needs to approve the workflow run first. Let's get it green before merging; the new fingerprint path and its regression tests need to be verified on CI, not just locally.

2. **Cover BOTH unreadable↔readable transitions, not just one.** Ansh's point 3 is about the `warm-then-becomes-unreadable` test. The mirror case matters too: a file that was **unreadable at warm time** (stored `fingerprint = None`) and later **becomes readable** must *invalidate* (`None` vs a real hash), while a still-unreadable file must *reuse* (`None` vs `None`). Please make sure the equality logic handles both directions and add a test for the unreadable→readable transition — that's the one most likely to silently serve stale records.

Not merging yet — for the maintainers once CI is green and the above + Ansh's points are addressed.

**Review by anilmurty — CHANGES_REQUESTED** [review]
noted in the comments

**Comment by anilmurty** [comment]
Thanks @Vam-si-krish. The whole-file SHA-256 fingerprint is the right direction and it does fix the middle-rewrite case. But the last round of feedback isn't addressed yet — the last code commit (`afc95df`) predates both reviews, and the only commit since is a bare `main` merge. Still open:

1. **Store-time divergence (Greptile P1):** `_fingerprint(path)` and `_parse_records(path)` are two separate reads; a rewrite landing between them stores version-A's fingerprint with version-B's records, so a later lookup of version-A serves version-B's stale records. Re-stat + re-fingerprint after the parse, and skip the store if anything moved.
2. **Lazy the fingerprint:** it currently hashes the whole file on every call before `_load`. Move it inside the hit check, after size+mtime already match — otherwise every cache miss hashes and then re-reads the same file.
3. `test_unreadable_file_still_reuses_its_cache_entry` patches `_fingerprint → None` at warm time, so it never exercises the warm-then-becomes-unreadable transition it claims to cover.
4. Add the **unreadable → readable** transition test (stored `None` fingerprint vs a real hash must invalidate) — that's the direction most likely to silently serve stale records.
5. Refresh the PR body — it still describes the abandoned first/last-4096-byte design and claims the base suite was failing (it's green — 3643 passed).

Also the branch is behind `main` — `gh pr update-branch` (or rebase) when you push the fixes. Ping me when it's ready for another look.

**Comment by anilmurty** [comment]
@Vam-si-krish - are you planning to work on this?

File: tokenjam/core/transcript_cache.py [hunk 1]
```diff
@@ -85,20 +124,23 @@ def cached_read_records(path: Path, cache_dir: Path) -> list[dict[str, Any]]:
     except OSError:
         return []
     size, mtime = st.st_size, st.st_mtime
+    fingerprint = _fingerprint(path)
```
[the remark this answers is not in the document]
Comment (Vam-si-krish) [reply tokenjam/core/transcript_cache.py]: Addressed in c0aa81b. The cache now re-stats and re-fingerprints after parsing and skips the store if metadata or the previously observed fingerprint changed. The regression test rewrites the source during parsing and verifies the divergent records are returned for that call but never persisted.

**Comment by Vam-si-krish** [comment]
@anilmurty @anshss — yes, I’ve addressed the review gate and pushed the update.

- fingerprinting is now lazy after the cached `size`/`mtime` match
- the source is re-statted and re-fingerprinted after parsing; divergent parses are not stored
- the readable → unreadable test now warms with a real fingerprint first
- unreadable → readable invalidation is covered
- the module docstring and PR body now describe the whole-file hash and read-vs-parse tradeoff
- the branch is updated to current `main`

Local verification:
- transcript-cache tests: 18 passed
- Ruff: passed
- mypy: 268 source files, no issues
- full non-e2e suite: 5,062 passed, 7 skipped; the six unrelated failures reproduce identically on clean `origin/main`

The refreshed CI run is waiting for maintainer approval: https://github.com/Metabuilder-Labs/tokenjam/actions/runs/30514071537

Once that workflow is approved, this is ready for another review.

**Review by anilmurty — APPROVED** [review]
Thanks @Vam-si-krish — this is a clean, well-reasoned fix. Going full-file SHA-256 instead of the edge-sampling I'd suggested in the issue was the right call: edge-sampling would've missed a middle-only rewrite, and your tests cover exactly that alongside the equal-size/same-mtime case, unreadable files, and legacy cache entries without a fingerprint. The deferred-fingerprint-on-cold-miss tradeoff is sensible and documented, and re-fingerprinting after the parse to refuse a divergent entry is a nice touch.

Addresses my earlier review. Merging. Thanks to @anshss for the co-author work too.
